### ISSUE TITLE
Support for bit precision in the Inference API text_embedding task

### ISSUE THREAD

**Issue by jimczi** [elastic/elasticsearch#111747]
### Description

Some inference API providers now support embedding models with each dimension defined as a single bit. For example, the v3 models from Cohere offer this capability. Since we already handle the bit element type in the dense vector field, it would be beneficial to extend this support to allow the text_embedding task of the inference API to output vectors with bit precision.

Typically, bit vectors are paired with float or byte vectors to improve recall by rescoring the hits from bit vectors with higher precision vectors. To support this, we suggest allowing the text_embedding task to generate multiple vectors for the same input at different precisions (e.g., bits + floats or bits + int8). While this functionality is already available in the Cohere API, implementing it in the inference API would optimize performance by eliminating the need to make two separate API calls for each precision, thereby reducing costs for users.

This would require the mapping to be defined with two fields, each corresponding to a different precision.

Additionally, we should evaluate whether the semantic_text field could natively support this scenario.

**Comment by ymao1** [comment]
> Some inference API providers now support embedding models with each dimension defined as a single bit. For example, the v3 models from Cohere offer this capability. Since we already handle the bit element type in the dense vector field, it would be beneficial to extend this support to allow the text_embedding task of the inference API to output vectors with bit precision.

Draft PR for adding bit precision support to Cohere text_embedding task: https://github.com/elastic/elasticsearch/pull/120751

With this PR, I was able to adapt the instructions from this blog post https://www.elastic.co/search-labs/blog/elasticsearch-cohere-embeddings-support to create two cohere models and an inference pipeline with two stages in order to generate text embeddings at different precisions.

Model for float embeddings
```
PUT _inference/text_embedding/cohere_embeddings_float
{
    "service": "cohere",
    "service_settings": {
        "api_key": <api-key>, 
        "model_id": "embed-english-v3.0", 
        "embedding_type": "float"
    }
}
```

Model for binary embeddings
```
PUT _inference/text_embedding/cohere_embeddings_binary
{
    "service": "cohere",
    "service_settings": {
        "api_key": <api-key>, 
        "model_id": "embed-english-v3.0", 
        "embedding_type": "binary"
    }
}
```

Index mappings
```
PUT cohere-embeddings
{
  "mappings": {
    "properties": {
      "name_embedding_float": { 
        "type": "dense_vector", 
        "dims": 1024, 
        "element_type": "float"
      },
      "name_embedding_binary": { 
        "type": "dense_vector", 
        "dims": 1024, 
        "element_type": "bit"
      },
      "name": { 
        "type": "text" 
      }
    }
  }
}
```

Ingest pipeline
```
PUT _ingest/pipeline/cohere_embeddings
{
  "processors": [
    {
      "inference": {
        "model_id": "cohere_embeddings_float", 
        "input_output": { 
          "input_field": "name",
          "output_field": "name_embedding_float"
        }
      }
    },
    {
      "inference": {
        "model_id": "cohere_embeddings_binary", 
        "input_output": { 
          "input_field": "name",
          "output_field": "name_embedding_binary"
        }
      }
    }
  ]
}
```

If we want to support allowing the text embedding task to generate multiple vectors at different precisions, here are some thoughts:

For updating the inference API input, we could:
1. Deprecate the `embeddings_type` option in the Cohere service settings in favor of a `embeddings_types` input that takes an array of embedding types that will be passed to the Cohere API
2. Add an optional `embedding_types` param to the Cohere task settings that overrides the `embedding_type` config in the service setting.

For updating the inference API output, currently we return responses in different keys depending on the embedding type, so it would be straightforward to update the response to return embeddings for multiple embedding types.

for `float` embedding type
```
{
  "text_embedding": [
    { "embedding": [....]}
  ]
}
```

for `byte` embedding type
```
{
  "text_embedding_bytes": [
    { "embedding": [....]}
  ]
}
```

for `binary` embedding type
```
{
  "text_embedding_bits": [
    { "embedding": [....]}
  ]
}
```

With this update to the output, it seems that we would also have to update the inference processor to handle multiple embedding type output and write to multiple output fields.

cc @davidkyle

**Comment by davidkyle** [comment]
Great example of generating embeddings at different precisions for the same input! This would be possible once https://github.com/elastic/elasticsearch/pull/120751 is merged. 

It does make sense to support requesting multiple `embedding_types` in a single API call but I'm not sure what it would look like for semantic_text. That is no reason not to merge #120751 as we need binary embedding support anyway. 

> For updating the inference API input, we could:
> 1. Deprecate the embeddings_type option in the Cohere service settings in favor of a embeddings_types input that takes an array of embedding types that will be passed to the Cohere API
> 2 Add an optional embedding_types param to the Cohere task settings that overrides the embedding_type config in the service setting.

The second option introduces another setting and a different place to configure that setting which can be confusing. 1. would be the better in my opinion.

I'd like to avoid deprecating `embeddings_type` singular. The request object could be changed so that it accepts either a single item or a list e.g. `embeddings_type: float` would be equivalent to `embeddings_type: [float]`. Elasticsearch has builtin support for parsing both forms and will automatically convert the single item into a list of 1 item. `embeddings_types` (plural) could be added as an alias for `embeddings_type`.

**Comment by ymao1** [comment]
> 1. would be the better in my opinion.

> I'd like to avoid deprecating embeddings_type singular. The request object could be changed so that it accepts either a single item or a list e.g. embeddings_type: float would be equivalent to embeddings_type: [float]. Elasticsearch has builtin support for parsing both forms and will automatically convert the single item into a list of 1 item. embeddings_types (plural) could be added as an alias for embeddings_type.

Makes sense! I can work on a POC for this and see what, if any, changes we need to the inference processor to support multi embedding type outputs.

**Comment by ymao1** [comment]
Added support for [Jina AI](https://github.com/elastic/elasticsearch/pull/121548) and [Cohere](https://github.com/elastic/elasticsearch/pull/120751)

**Comment by jonathan-buttner** [comment]
~@jimczi, weirdly I'm not able to close this. I think we've implemented this functionality. Could you close the issue?~

JK, I found a way!
